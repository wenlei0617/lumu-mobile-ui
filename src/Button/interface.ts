export interface ButtonProps {
  /**
   * @description 按钮类型，可选值 default primary
   * @default 'default'
   */
  type?: 'default' | 'primary';
  /**
   * @description 设置大小，可选值 md lg sm
   * @default 'md'
   */
  size?: 'md' | 'lg' | 'sm';
  /**
   * @description 设置形状，可选值为 rect radius round
   * @default 'radius''
   */
  shape?: 'rect' | 'radius' | 'round';
  /**
   * @description 是否块级元素
   * @default false
   */
  block?: boolean;
  /**
   * @description 是否禁用
   * @default false
   */
  disabled?: boolean;
  /**
   * @description 是否加载中状态
   * @default false
   */
  loading?: boolean;
  /**
   * @description 设置图标
   */
  icon?: React.ReactNode;
  /**
   * @description 点击后触发的回调函数
   */
  onClick?: React.MouseEventHandler<Element>;
}