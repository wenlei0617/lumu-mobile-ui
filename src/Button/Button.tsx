import { getClsPrefix } from '../utils/utils';
import React, { useCallback, useMemo } from 'react';
import classNames from 'classnames';
import { ButtonProps } from './interface';
import Icon from '../icon/index';
import './style/index.less';

const btnPrefix = getClsPrefix('btn');

export const Button: React.FC<ButtonProps> = (props) => {
  const {
    size = 'md',
    shape = 'radius',
    block = false,
    type = 'default',
    icon,
    loading = false,
    disabled = false,
    onClick,
    children,
  } = props;

  const className = useMemo(() => {
    return classNames(
      btnPrefix,
      `${btnPrefix}-size__${size}`,
      `${btnPrefix}-theme__${type}`,
      `${btnPrefix}-shape__${shape}`,
      {
        [`${btnPrefix}-block`]: block,
        [`${btnPrefix}-disabled`]: disabled,
      },
    );
  }, [size, type, block, shape, disabled]);

  const handleClick = useCallback(
    (e) => {
      if (loading || disabled) return;
      onClick?.(e);
    },
    [onClick, loading, disabled],
  );

  return (
    <button type="button" onClick={handleClick} className={className}>
      <div className={`${btnPrefix}-content`}>
        {loading && (
          <Icon type="icon-loading" className={`${btnPrefix}-loading`} />
        )}
        {icon && !loading && (
          <span className={`${btnPrefix}-icon`}>{icon}</span>
        )}
        <span className={`${btnPrefix}-text`}>{children}</span>
      </div>
    </button>
  );
};
