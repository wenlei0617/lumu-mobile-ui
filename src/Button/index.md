## Button

按使用场景：
- 普通按钮
- 图标按钮
- 文本按钮

按大小：
- 小按钮
- 标准按钮
- 大按钮

按形状：
- 矩形
- 弧形
- 圆形

按主题色：
- 主题按钮

按空间：
- 块级元素
- 行类元素

按状态：
- 默认状态
- 禁止状态
- 加载状态

简单用法:
<code src="./demo/index.tsx"></code>

<API></API>
