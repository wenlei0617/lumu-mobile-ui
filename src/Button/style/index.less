@import '../../style/common.less';
@import '//at.alicdn.com/t/font_2921882_chtev9wf9u.css';

@btn-prefix-cls: ~'@{lumu-prefix}-btn';

.theme(@color: @theme) {
  border: 1px solid @color;
  color: @color;
}

.@{btn-prefix-cls} {
  background-color: #fff;
  border: none;
  display: inline-block;
  user-select: none;
  cursor: pointer;
  touch-action: manipulation;
  font-weight: normal;
  box-sizing: border-box;
  white-space: nowrap;
  overflow: hidden;
  outline: 0;
  position: relative;
  -webkit-appearance: none;
  -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
  &::after {
    content: '';
    position: absolute;
    width: 100%;
    height: 100%;
    top: 50%;
    left: 50%;
    background-color: #000;
    border: inherit;
    border-color: #000;
    border-radius: inherit;
    transform: translate(-50%, -50%);
    opacity: 0;
    transition: opacity 300ms linear;
  }
  &:active::after {
    opacity: .1;
  }
  &-text {
    overflow: hidden;
    text-overflow: ellipsis;
  }
  &-loading {
    margin-right: 6px;
    animation: rotate linear 1s infinite;
    font-size: inherit;
    flex-shrink: 0;
  }
  @keyframes rotate {
    0% {
      transform: rotate(0deg);
    }
    100% {
      transform: rotate(360deg);
    }
  }
  &-content {
    display: flex;
    width: 100%;
    height: 100%;
    justify-content: center;
    align-items: center;
    overflow: hidden;
  }
  &-size {
    &__sm {
      height: 28px;
      padding: 0 12px;
      font-size: 12px;
    }
    &__md {
      height: 38px;
      padding: 0 18px;
      font-size: 14px;
    }
    &__lg {
      height: 48px;
      padding: 0 24px;
      font-size: 14px;
    }
  }
  &-block {
    display: block;
    width: 100%;
  }
  &-theme {
    &__default {
      border: 1px solid @theme-gray;
      color: @theme-gray;
    }
    &__primary {
      border: 1px solid @theme;
      color: #fff;
      background-color: @theme;
    }
  }
  &-shape {
    &__rect {
      border-radius: 0;
    }
    &__radius {
      border-radius: 5px;
    }
    &__round {
      border-radius: 18px;
    }
  }
  &-disabled {
    background-color: @theme-gray;
    color: #fff;
    cursor: not-allowed;
    border: none;
  }
}
