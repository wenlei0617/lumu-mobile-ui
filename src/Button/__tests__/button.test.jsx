import React from 'react';
import Button from '../index';
import { render, fireEvent, screen } from '@testing-library/react';

describe('Button', () => {
  test('正常挂载', async () => {
    const { getByText } = render(<Button>普通按钮</Button>);
    expect(getByText('普通按钮')).toBeInTheDocument();
  });

  test('普通按钮', async () => {
    const { asFragment } = render(<Button>普通按钮</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  test('图标按钮', async () => {
    const { asFragment } = render(<Button icon="😀">图标按钮</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  test('小按钮', async () => {
    const { asFragment } = render(<Button size="sm">小按钮</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  test('中按钮', async () => {
    const { asFragment } = render(<Button size="md">中按钮</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  test('大按钮', async () => {
    const { asFragment } = render(<Button size="lg">大按钮</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  test('矩形按钮', async () => {
    const { asFragment } = render(<Button shape="rect">矩形按钮</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  test('圆形按钮', async () => {
    const { asFragment } = render(<Button shape="round">圆形按钮</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  test('块级按钮', async () => {
    const { asFragment } = render(<Button block>块级按钮</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  test('禁止按钮', async () => {
    const { asFragment } = render(<Button disabled>禁止按钮</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  test('加载按钮', async () => {
    const { asFragment } = render(<Button loading>加载按钮</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  test('主题按钮', async () => {
    const { asFragment } = render(<Button type="primary">主题按钮</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  test('点击按钮', async () => {
    const handleClick = jest.fn();
    render(<Button onClick={handleClick}>点击按钮</Button>);
    fireEvent.click(screen.getByText('点击按钮'));
    expect(handleClick).toHaveBeenCalledTimes(1);
  });
});
