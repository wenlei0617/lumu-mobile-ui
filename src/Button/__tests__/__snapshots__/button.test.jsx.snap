// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button 中按钮 1`] = `
<DocumentFragment>
  <button
    class="lumu-btn lumu-btn-size__md lumu-btn-theme__default lumu-btn-shape__radius"
    type="button"
  >
    <div
      class="lumu-btn-content"
    >
      <span
        class="lumu-btn-text"
      >
        中按钮
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button 主题按钮 1`] = `
<DocumentFragment>
  <button
    class="lumu-btn lumu-btn-size__md lumu-btn-theme__primary lumu-btn-shape__radius"
    type="button"
  >
    <div
      class="lumu-btn-content"
    >
      <span
        class="lumu-btn-text"
      >
        主题按钮
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button 加载按钮 1`] = `
<DocumentFragment>
  <button
    class="lumu-btn lumu-btn-size__md lumu-btn-theme__default lumu-btn-shape__radius"
    type="button"
  >
    <div
      class="lumu-btn-content"
    >
      <span
        class="iconfont icon-loading lumu-btn-loading"
      />
      <span
        class="lumu-btn-text"
      >
        加载按钮
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button 图标按钮 1`] = `
<DocumentFragment>
  <button
    class="lumu-btn lumu-btn-size__md lumu-btn-theme__default lumu-btn-shape__radius"
    type="button"
  >
    <div
      class="lumu-btn-content"
    >
      <span
        class="lumu-btn-icon"
      >
        😀
      </span>
      <span
        class="lumu-btn-text"
      >
        图标按钮
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button 圆形按钮 1`] = `
<DocumentFragment>
  <button
    class="lumu-btn lumu-btn-size__md lumu-btn-theme__default lumu-btn-shape__round"
    type="button"
  >
    <div
      class="lumu-btn-content"
    >
      <span
        class="lumu-btn-text"
      >
        圆形按钮
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button 块级按钮 1`] = `
<DocumentFragment>
  <button
    class="lumu-btn lumu-btn-size__md lumu-btn-theme__default lumu-btn-shape__radius lumu-btn-block"
    type="button"
  >
    <div
      class="lumu-btn-content"
    >
      <span
        class="lumu-btn-text"
      >
        块级按钮
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button 大按钮 1`] = `
<DocumentFragment>
  <button
    class="lumu-btn lumu-btn-size__lg lumu-btn-theme__default lumu-btn-shape__radius"
    type="button"
  >
    <div
      class="lumu-btn-content"
    >
      <span
        class="lumu-btn-text"
      >
        大按钮
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button 小按钮 1`] = `
<DocumentFragment>
  <button
    class="lumu-btn lumu-btn-size__sm lumu-btn-theme__default lumu-btn-shape__radius"
    type="button"
  >
    <div
      class="lumu-btn-content"
    >
      <span
        class="lumu-btn-text"
      >
        小按钮
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button 普通按钮 1`] = `
<DocumentFragment>
  <button
    class="lumu-btn lumu-btn-size__md lumu-btn-theme__default lumu-btn-shape__radius"
    type="button"
  >
    <div
      class="lumu-btn-content"
    >
      <span
        class="lumu-btn-text"
      >
        普通按钮
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button 矩形按钮 1`] = `
<DocumentFragment>
  <button
    class="lumu-btn lumu-btn-size__md lumu-btn-theme__default lumu-btn-shape__rect"
    type="button"
  >
    <div
      class="lumu-btn-content"
    >
      <span
        class="lumu-btn-text"
      >
        矩形按钮
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button 禁止按钮 1`] = `
<DocumentFragment>
  <button
    class="lumu-btn lumu-btn-size__md lumu-btn-theme__default lumu-btn-shape__radius lumu-btn-disabled"
    type="button"
  >
    <div
      class="lumu-btn-content"
    >
      <span
        class="lumu-btn-text"
      >
        禁止按钮
      </span>
    </div>
  </button>
</DocumentFragment>
`;
